crates/proto/src/http/mod.rs

```diff
@@ -7,17 +7,159 @@
 
 //! HTTP protocol related components for DNS over HTTP/2 (DoH) and HTTP/3 (DoH3)
 
+use alloc::sync::Arc;
 use core::num::ParseIntError;
+use core::str::FromStr;
 use std::io;
 
 use alloc::string::String;
-use http::header::{CONTENT_LENGTH, CONTENT_TYPE, ToStrError};
-use http::{HeaderMap, HeaderValue, Response, StatusCode};
+use http::header::{ACCEPT, CONTENT_LENGTH, CONTENT_TYPE, ToStrError};
+use http::{HeaderMap, HeaderValue, Request, Response, StatusCode, Uri, header, uri};
 use thiserror::Error;
+use tracing::debug;
 
 use crate::error::ProtoError;
 
-pub mod request;
+pub(crate) struct RequestContext {
+    pub(crate) version: Version,
+    pub(crate) server_name: Arc<str>,
+    pub(crate) query_path: Arc<str>,
+    pub(crate) set_headers: Option<Arc<dyn SetHeaders>>,
+}
+
+impl RequestContext {
+    /// Create a new Request for an http dns-message request
+    ///
+    /// ```text
+    /// RFC 8484              DNS Queries over HTTPS (DoH)          October 2018
+    ///
+    /// The URI Template defined in this document is processed without any
+    /// variables when the HTTP method is POST.  When the HTTP method is GET,
+    /// the single variable "dns" is defined as the content of the DNS
+    /// request (as described in Section 6), encoded with base64url
+    /// [RFC4648].
+    /// ```
+    pub(crate) fn build(&self, message_len: usize) -> Result<Request<()>, Error> {
+        let mut parts = uri::Parts::default();
+        parts.path_and_query = Some(
+            uri::PathAndQuery::try_from(&*self.query_path)
+                .map_err(|e| ProtoError::from(format!("invalid DoH path: {e}")))?,
+        );
+        parts.scheme = Some(uri::Scheme::HTTPS);
+        parts.authority = Some(
+            uri::Authority::from_str(&self.server_name)
+                .map_err(|e| ProtoError::from(format!("invalid authority: {e}")))?,
+        );
+
+        let url = Uri::from_parts(parts)
+            .map_err(|e| ProtoError::from(format!("uri parse error: {e}")))?;
+
+        // TODO: add user agent to TypedHeaders
+        let mut request = Request::builder()
+            .method("POST")
+            .uri(url)
+            .version(self.version.to_http())
+            .header(CONTENT_TYPE, crate::http::MIME_APPLICATION_DNS)
+            .header(ACCEPT, crate::http::MIME_APPLICATION_DNS)
+            .header(CONTENT_LENGTH, message_len);
+
+        if let Some(headers) = &self.set_headers {
+            if let Some(map) = request.headers_mut() {
+                headers.set_headers(map)?;
+            }
+        }
+
+        Ok(request
+            .body(())
+            .map_err(|e| ProtoError::from(format!("http stream errored: {e}")))?)
+    }
+}
+
+/// Verifies the request is something we know what to deal with
```

Comment:
pre-existing, but I think this comment reads oddly. For the most minimal change, maybe something like:

> Verifies the request is something we know how to deal with

Even better would be to describe some of what it means to know how to deal with it, maybe something like:

> Verifies the request is well-formed for the name-server and supported protocols